### PR TITLE
Implement LoggingAsync for GRPC Worker Services

### PR DESCRIPTION
**Pull request by xldrx**
This allow RecvTensor events to show up in StepStats and in turn in Chrome Tracing format.

### CODE REVIEW

**Comment by xldrx** [comment]
@pbar @mrry @suharshs Could you please take a look at this?

**Comment by mrry** [comment]
Reassigning this to @prb12, since he knows the ins and outs of the logging path better than I do. A few quick observations though:

* Instead of modifying the `BaseRendezvousMgr` and related classes to capture this information, please use the existing stubs in `GrpcWorkerCache`:

  https://github.com/tensorflow/tensorflow/blob/c0662f1620c2b97abb79b8ae6a8a30f7c7719475/tensorflow/core/distributed_runtime/rpc/grpc_worker_cache.cc#L86-L92

* The changes in `master_session.cc` are potentially controversial, because they could cause a lot of additional logging to be generated when `FULL_TRACE` is specified. We might need an additional `TraceLevel` for this case.

* Run `clang-format -style=Google` over the code to take care of style issues (which should trigger a failure in the presubmit).

**Comment by xldrx** [comment]
Thanks @mrry for the feed back:
* I changed the implementation to use SessionMgr instead of *RendezvousMgr. 
Note that unlike other worker rpcs, `LoggingRequest` in  `LoggingAsync` rpc does not have any session handle, therefore all sessions should be touched. Both implementations (previous one, and this one) iterates over all sessions (previously through `RendezvousMgr` now `SessionMgr`), then uses `GrpcWorkerCache` calls in each session.

* @pbar: Thoughts on changes in `master_session.cc`?

**Comment by sb2nov** [comment]
@xldrx can you look into the merge conflicts.

**Comment by xldrx** [comment]
Hey @sb2nov,
It seems that Paul is swamped. Could you review this PR?

**Comment by sb2nov** [comment]
@xldrx I don't know enough to review this right now. @prb12 ping.

File: tensorflow/core/distributed_runtime/rpc/grpc_worker_service.cc [hunk 1]
```diff
@@ -392,6 +392,24 @@ void GrpcWorker::GrpcRecvTensorAsync(CallOptions* opts,
...
+void GrpcWorker::LoggingAsync(const LoggingRequest* request,
+                              LoggingResponse* response, StatusCallback done) {
+  auto env = this->env();
+  if (env) {
+    auto session_mgr = (SessionMgr*)env->session_mgr;
+    if (session_mgr) {
+      session_mgr->SetLogging(request->rpc_logging());
+      if (request->clear()) {
+        session_mgr->ClearLogs();
+      }
```
Comment (prb12) [inline tensorflow/core/distributed_runtime/rpc/grpc_worker_service.cc]: If both `clear` and `fetch_step_id` are used the semantics are unclear, but I would have thought it made more sense to retrieve any previously collected data before clearing?

**Review by prb12 — CHANGES_REQUESTED** [review]
Thanks for addiing this!

With regard to an additional TRACE_LEVEL - I don't think that this is likely to be a problem.  

I'm assuming this doesn't have a noticeable performance impact when tracing is not enabled?

One minor question about the case where clear and retrieve are done at the same time.

File: tensorflow/core/distributed_runtime/rpc/grpc_worker_service.cc (hunk 1 above)
Comment (xldrx) [reply tensorflow/core/distributed_runtime/rpc/grpc_worker_service.cc]: Great observation. Fixed.

**Comment by xldrx** [comment]
@prb12 Thanks for the review. 
> I'm assuming this doesn't have a noticeable performance impact when tracing is not enabled?

This PR does nothing when tracing is not enabled.

> One minor question about the case where clear and retrieve are done at the same time.

Great observation. Fixed.

**Review by prb12 — APPROVED** [review]

**Comment by rmlarsen** [comment]
@xldrx This appears to have numerous broken tests. Can you please take a look?

**Review by prb12 — APPROVED** [review]

**Comment by rmlarsen** [comment]
@xldrx Thanks for the contribution!

**Pull request merged**